video/out/opengl/context_drm_egl.c

```diff
@@ -319,12 +319,15 @@ static void queue_flip(struct ra_ctx *ctx, struct gbm_frame *frame)
     update_framebuffer_from_bo(ctx, frame->bo);
 
     struct drm_atomic_context *atomic_ctx = drm->atomic_context;
+    int flags = DRM_MODE_ATOMIC_NONBLOCK | DRM_MODE_PAGE_FLIP_EVENT;
     drm_object_set_property(atomic_ctx->request, atomic_ctx->draw_plane, "FB_ID", drm->fb->id);
     drm_object_set_property(atomic_ctx->request, atomic_ctx->draw_plane, "CRTC_ID", atomic_ctx->crtc->id);
     drm_object_set_property(atomic_ctx->request, atomic_ctx->draw_plane, "ZPOS", 1);
 
-    int ret = drmModeAtomicCommit(drm->fd, atomic_ctx->request,
-                                  DRM_MODE_ATOMIC_NONBLOCK | DRM_MODE_PAGE_FLIP_EVENT, drm);
+    if (vo_drm_set_hdr_metadata(ctx->vo, false))
```

Comment:
Not really. Not sending color information will produce incorrect output on display.

Reply:
It should unset itself if mpv is terminated properly. Of course we can't do anything about a SIGKILL.

I can make this respect `--target-colorspace-hint` but probably we should just set that option to true by default I think.

Reply:
There is misconception about what `--target-colorspace-hint` does. It is not meant to control signaling to the display, it is meant to automatically set `--target-` if possible, signaling colorspace should be done unconditionally. Note that it controls how video is rendered to ensure it always is correct. Skipping signaling part in drm layer is too late to do. 

Also I already said multiple times there has to be negotiation between output layer and gpu-next.c to negotiate correct target-colorspace-hint behavior. It may resolve your concern. Currently this PR is impossible to use, because you have to manually set each and all HDR parameters. And even so, it is currently not possible to pass HDR metadata. Simply because how gpu-next is wired up, `--target-trc=pq` will use 0/10000 HDR10 metadata. Only way is to support this property.

So in fact I agree `--target-colorspace-hint` has to be supported/fixed, not in drm.c, but in gpu-next.c. As shortly described in https://github.com/mpv-player/mpv/pull/15048#discussion_r1798704514

> The current solution only works with a limited set of color gamut and features

So it won't do anything for others, nothing breaks.

> and there is no way to disable this when the hardware or driver is buggy

It is not a mpv job to workaround buggy drivers. And for sure not something we should do preemptively. If this is a problem, we can diagnose the root cause and think of the solution that fits certain case.

> or if I prefer mpv's internal tonemapping

What do you mean? This is precisely what it does now. You can set `--target-` to your liking tonemapping or not and it would signal proper image characteristics based on target_params.

> The display will also be stuck in HDR mode if mpv is killed.

Nothing we can do about it if we get SIGKILL and how this is related to this discussion?

Reply:
It's worth noting that AFAIK there's no way for us to know what the right `target-prim` is on this level unless anybody has any ideas.

Reply:
I think there is some misunderstanding here. This doesn't unconditionally signal anything. By default it will always be SDR because that's what the target_params will be. You have to manually specify `--target-prim` and `--target-trc` for this to even do anything. It would be nice if `--target-colorspace-hint` worked, but I have no clue how you would be able to find the right prim for the underlying display. As some people testing in this very PR noted, some things needed bt.2020 and others needed display-p3.

A generic mechanism like kasper was talking about earlier to signal colorspaces/metadata might be nice but I don't think it would be useful for drm. At least not in its current state anyway. That would have more practical use for wayland.

Reply:
>My point is that just specifying --target-prim and --target-trc by themselves shouldn't signal the colorspace, because that's also the case for gpu-next when --target-colorspace-hint=no.

But they do?

Reply:
> My point is that just specifying `--target-prim` and `--target-trc` by themselves shouldn't signal the colorspace, because that's also the case for gpu-next when `--target-colorspace-hint=no`.

That's valid point, but truth is current system in gpu-next is bit broken, because it works only if the colorspace is signaled through libplacebo which works only with Vulkan and DXGI api.

This whole system needs to be refactored, now that we have ability to signal colorspace in meaningful way in mpv. For example currently there is no way to pass real metadata to DRM, sure it will set PQ in target_params, but max_luma/min_luma will be incorrect.

`--target-colorspace-hint=no` with `--target-trc=pq` doesn't make any sense, except maybe for debugging, but this should never be needed. Depending on API used, there are multiple places where the colorspace needs to be set correctly. You cannot set both of those options and expect it to display correctly. Even if your screen is in "HDR" mode, there are thing in-between that might just change that if colorspace is not "set".

It is as if we had `--target-resolution` and `--target-resolution-hint=no` and try to copy 1080p buffer onto 720p surface... 

There has to be implemented negotiation between drm and gpu-next, what is supported and what should be on output. `--target-colorspace-hint=no` can be left for debugging purposes, but I would rename it to `--target-colorspace=<auto|no>`. Where auto would do thing properly and `no` would be SDR fallback for whatever reason. We live in a word where not everything connected to PC is sRGB anymore and this has to be reflected.

Note that vo_gpu is unconditionally signaling HDR with D3D11, but I don't want to go into vo_gpu territory, it is another level of complexity to solve...

> But they do?

They won't signal anything with `--target-colorspace-hint=no`, resulting in broken colors.

Reply:
It's overriden in `apply_target_options`. Calling `pl_swapchain_colorspace_hint` is just what lets libplacebo infer the colorspace or not which is not really relevant here.

Reply:
> It's overriden in `apply_target_options`. Calling `pl_swapchain_colorspace_hint` is just what lets libplacebo infer the colorspace or not which is not really relevant here.

It is relevant, because without it swapchain will not be reconfigured properly for HDR content. It will just be srgb default one.

Reply:
> Well I meant for DRM egl though. There is nothing extra libplacebo side should be doing here AFAIK. It completely depends on the drm code.

gpu_next sets teraget_params to properties of the swapchain, so it won't set HDR if colorspace_hint is not enabled/working. So it doesn't work correctly at this moment in time.